R/join_keys.R

```diff
@@ -0,0 +1,651 @@
+#' is not empty.
+#'
+#' @param join_keys_obj (`join_keys`) empty object to set the new relationship pairs.
+#' @param value (`join_key_set` or list of `join_key_set`) relationship pairs to add
+#' to `join_keys` list.
+#'
+#' @export
+`join_keys<-` <- function(join_keys_obj, value) {
+  UseMethod("join_keys<-", join_keys_obj)
+}
+
+#' @rdname join_keys
+#' @export
+#' @examples
+#'
+#' # Using the setter (assignment) ----
+#'
+#' jk <- join_keys()
+#' join_keys(jk) <- join_key("ds1", "ds2", "some_col")
+#' join_keys(jk) <- c(join_keys(jk), join_key("ds3", "ds4", "some_col2"))
+#'
+#' join_keys(jk)[["ds1"]][["ds3"]] <- "some_col3"
+#' jk
+`join_keys<-.join_keys` <- function(join_keys_obj, value) {
+  if (missing(value)) {
+    return(join_keys_obj)
+  }
+
+  # Assume assignment of join keys as a merge operation
+  #  Needed to support join_keys(jk)[ds1, ds2] <- "key"
+  if (checkmate::test_class(value, classes = c("join_keys", "list"))) {
+    return(value)
+  }
+
+  if (inherits(value, "join_key_set")) value <- list(value)
+
+  checkmate::assert_list(value, types = "join_key_set", min.len = 1)
+
+  join_keys_obj <- new_join_keys()
+
+  # check if any join_key_sets share the same datasets but different values
+  for (idx_1 in seq_along(value)) {
+    for (idx_2 in seq_along(value[idx_1])) {
+      assert_compatible_keys(value[[idx_1]], value[[idx_2]])
+    }
+    dataset_1 <- get_dataset_1(value[[idx_1]])
+    dataset_2 <- get_dataset_2(value[[idx_1]])
+    keys <- get_keys(value[[idx_1]])
+
+    join_keys_obj[[dataset_1]][[dataset_2]] <- keys
+  }
+
+  logger::log_trace("join_keys keys are set.")
+
+  join_keys_obj
+}
+
+#' @rdname join_keys
+#' @export
+#'
+#' @examples
+#'
+#' c(join_keys(join_key("a", "b", "c")), join_keys(join_key("a", "d2", "c")))
+c.join_keys <- function(...) {
+  x <- rlang::list2(...)
+
+  if (!length(x)) {
+    return(NULL)
+  }
+  checkmate::assert_list(x[-1], types = c("join_keys", "join_key_set"))
+
+  merge_join_keys.join_keys(x[[1]], x[-1])
+}
+
+#' @rdname join_keys
+#' @export
+#' @examples
+#'
+#' # Setter for join_keys within teal_data ----
+#'
+#' td <- teal_data()
+#' join_keys(td)[["ds1"]][["ds2"]] <- "key1"
+#' join_keys(td)[["ds2"]][["ds2"]] <- "key2"
+#' join_keys(td) <- c(join_keys(td), join_keys(join_key("ds3", "ds2", "key3")))
+#' join_keys(td)
+`join_keys<-.teal_data` <- function(join_keys_obj, value) {
+  if (missing(value)) {
+    return(join_keys_obj)
+  }
+
+  if (checkmate::test_class(value, c("join_keys", "list"))) {
+    join_keys_obj@join_keys <- merge_join_keys(join_keys_obj@join_keys, value)
+    return(join_keys_obj)
+  }
+
+  join_keys(join_keys_obj@join_keys) <- value
+  join_keys_obj
+}
+
+#' The Names of an `join_keys` Object
+#' @inheritParams base::`names<-`
+#' @export
+`names<-.join_keys` <- function(x, value) {
+  x <- unclass(x)
+  # Update inner keys
+  for (old_name in setdiff(names(x), value)) {
+    old_entry <- x[[old_name]]
+    new_name <- value[names(x) == old_name]
```

Comment:
This looks a lot like `repalce(names(x), setdiff(names(x), value), value)`.

Reply:
I don't think this can be done as `value` on `*<-` functions has always the same length as `x`. Even when addressed with indexes.

The `value` will have the unchanged names as well as the changed ones

```r
ex <- rlang::set_names(c("1", "2", "3", "4"))
names(ex)[1] <- "11"
```

Will internally do:

```r
new_change_in_names <- names(ex)
new_change_in_names[1] <- "11"
`names<-.class`(x = ex, value = new_change_in_names)
````

Reply:
All right. I thought it could be made a little less convoluted but if you tried out different solutions, I trust you picked the best one.